### ISSUE TITLE
Not possible to execute tests with parameterized TestFixtures

### ISSUE THREAD

**Issue by gustafj** [formulahendry/vscode-dotnet-test-explorer#144]
## Description

When trying to run a test case which has a [Parameterized TestFixture](https://github.com/nunit/docs/wiki/TestFixture-Attribute#parameterized-test-fixtures) the executions fails due to not finding the test case.
The reason is that the FQN used is missing the parameters.
This is the current output I get for the example below.
```
Executing dotnet test --logger "trx;LogFileName=/tmp/test-explorer/0.trx" --filter FullyQualifiedName~example.ExampleTests.Test_Example in ...
Build started, please wait...
Build completed.

Test run for .../example.dll(.NETCoreApp,Version=v2.0)
Microsoft (R) Test Execution Command Line Tool Version 15.3.0-preview-20170628-02
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
NUnit Adapter 3.10.0.21: Test execution started
Running all tests in .../example.dll
NUnit3TestExecutor converted 1 of 1 NUnit test cases
Skipping assembly - no matching test cases found
NUnit Adapter 3.10.0.21: Test execution complete
No test matches the given testcase filter `FullyQualifiedName~example.ExampleTests.Test_Example` in .../example.dll
```

## Solution

This is possible to execute on the command line using this filter:
```
dotnet test --logger "trx;LogFileName=/tmp/output.trx" --filter '"FullyQualifiedName~example.ExampleTests\(\"First\"\).Test_Example"'
```

## Example

```csharp
namepace example {
    [TestFixture("First")]
    [TestFixture("Second")]
    public class ExampleTests
    {
        public ExampleTests(string arg)
        {
            // Do something with arg that affects the tests
        }

        [TestCase]
        public void Test_Example()
        {
            // Test something
        }
    }
}
```

**Comment by stefanforsberg** [comment]
When I execute 
`dotnet test --logger "trx;LogFileName=/tmp/output.trx" --filter '"FullyQualifiedName~example.ExampleTests\(\"First\"\).Test_Example"'`

I receive an error "An exception occurred while invoking executor 'executor://nunit3testexecutor/': Incorrect format for TestCaseFilter Missing Operator '|' or '&'. Specify the correct format and try again. Note that the incorrect format can lead to no test getting executed."

My guess is that you are on mac perhaps? Or using some other shell on windows. 

The problem with implementing this feature is that the command may differ between test frameworks and on shell. Not being able to run a specific test fixture test is understandably a bummer but at least you can run _all_ of them consistently regardless of shell, os and framework. 

Also, we love pull requests =)

**Comment by gustafj** [comment]
I'm actually on linux, using the `fish` shell though, however I just tried with `bash` (only other shell I have installed) and it parsed successfully there as well.

I do get that error if I don't escape the parenthesis or have another syntax issue though, did something go wrong in a copy-paste?

I used this as reference for the syntax https://github.com/Microsoft/vstest-docs/blob/master/docs/filter.md

Thanks for the prompt response! :+1:

**Comment by stefanforsberg** [comment]
Usually I'm pretty good at copy/pasting 🥇 but I've tried and failed on both my pc and my mac now, I get the same error I got above.

I wonder if this is something that perhaps has changed in the .net tooling or in the adapter version of the test framework. What version of the dotnet sdk and what version of nunit are you using?

**Comment by gustafj** [comment]
Hehe, check :smile:  :1st_place_medal:  

```
> dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.401
 Commit:    91b1c13032

Runtime Environment:
 OS Name:     arch
 OS Version:  
 OS Platform: Linux
 RID:         arch-x64
 Base Path:   /opt/dotnet/sdk/2.1.401/

Host (useful for support):
  Version: 2.1.3
  Commit:  124038c13e

.NET Core SDKs installed:
  2.1.401 [/opt/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.NETCore.App 2.1.3 [/opt/dotnet/shared/Microsoft.NETCore.App]
```
```
<PackageReference Include="NUnit" Version="3.10.1" />
<PackageReference Include="NUnit3TestAdapter" Version="3.10.0" />
```
I only have one version of NUnit installed
Any other version info you need?

**Comment by stefanforsberg** [comment]
Do you have a reference to Microsoft.NET.Test.Sdk as well? We're currently using Version 15.5.0 in our test project. Could you share what version you have and if you have the time to try and run that command with the version we're using here?

**Comment by gustafj** [comment]
Yes, we do, and I tested with 15.5.0 which then fails on the same error as you get.
We are currently using 15.8.0 which works, I also tested 15.6.0, 15.7.0 & 15.7.2 as well, all fail. So it seems that the parsing was fixed / changed in 15.8.0.
`<PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.8.0" />`

The documentation of this syntax was added 8th of June [#141](https://github.com/Microsoft/vstest-docs/pull/141/files).
15.8.0 was released 13th of July [nuget](https://www.nuget.org/packages/Microsoft.NET.Test.Sdk/15.8.0).
15.7.2 was released 16th of May [nuget](https://www.nuget.org/packages/Microsoft.NET.Test.Sdk/15.7.2).

I don't know when the actual logic was added though.
Are you able to detect which version of _Microsoft.NET.Test.Sdk_ is used by the active project? And then if `version >= 15.8.0` invoke tests using the full FQN with arguments?

**Comment by stefanforsberg** [comment]
There is not an easy way to detect which version of a certain package is used that I'm aware of at least. We do some stuff in this area to get a rough idea which test framework people who are using our extension are using but it's super brittle and not something I'd like to base actual functionality on.

I guess we could add functionality that requires a certain min version of the Microsoft.NET.Test.Sdk and if people run into problem we can tell them to upgrade. Since you are the first one to report this issue I'm guessing it's not super common either and the solution with escaping should only affect people who actually needs to do the escaping (in this case using parameterized TestFixtures).

What are your thoughts on the matter @formulahendry?

**Comment by gustafj** [comment]
Any progress on this?
@formulahendry care to comment?
I'd like to help out in any way I can, but would need some pointers to where to start.

**Comment by stefanforsberg** [comment]
Not any real progress I'm afraid. It seem to fall smack in the middle of the hot mess that is different test frameworks behaving differently and different os:es behaving differently.

Current list of problems that's stopping the implementation are:

1. XUnit can not run a specific InlineData test, so those would break if we include the (escaped) test name (basically TestName("StuffThatGoesHere"))

2. Escaping the test name like you modeled above works fine with Microsoft.NET.Test.Sdk Version>="15.8.0... but not on windows. Windows really don't want to escape the quote char. It works fine with a test case with int but the string one I couldn't get to work.

``` csharp
[TestFixture(1)]
    public class ExampleTests
    {
        public ExampleTests(int arg)
        {
            // Do something with arg that affects the tests
        }
```

**Comment by gustafj** [comment]
Aww, that's just sad.

For the XUnit case, is it not known to the explorer if XUnit is used (as in `using Xunit`)?
If so that case should be solvable right?

For the second case, would a platform specific code path be acceptable?
